torch/lib/c10d/Utils.hpp

```diff
@@ -0,0 +1,196 @@
+  if ((port < 0) || (port >= std::numeric_limits<PortType>::max()))
+    throw std::domain_error("invalid port (value out of range)");
+
+  return static_cast<PortType>(port);
+}
+
+inline RankType convertToRank(int64_t rank, int64_t min = 0) {
+  if ((rank < min) || (rank >= std::numeric_limits<RankType>::max()))
+    throw std::domain_error("invalid rank (value out of range)");
+
+  return static_cast<RankType>(rank);
+}
+
+// TCP util namespace
+namespace tcputil {
+
+constexpr std::chrono::milliseconds kNoTimeout =
+  std::chrono::milliseconds(-1);
+
+// Send and receive
+template<typename T>
+void sendBytes(int socket,
+               const T* buffer,
+               size_t length,
+               bool moreData = false) {
+
+  size_t bytesToSend = sizeof(T) * length;
+  if (bytesToSend == 0) {
+    return;
+  }
+
+  auto bytes = reinterpret_cast<const uint8_t*>(buffer);
+  uint8_t* currentBytes = const_cast<uint8_t*>(bytes);
+
+  int flags = 0;
+
+#ifdef MSG_MORE
+  if (moreData) { // there is more data to send
+    flags |= MSG_MORE;
+  }
+#endif
+
+  while (bytesToSend > 0) {
+    ssize_t bytesSent;
+    SYSCHECK(bytesSent = ::send(socket, currentBytes, bytesToSend, flags))
+    if (bytesSent == 0) {
+      throw std::system_error(ECONNRESET, std::system_category());
+    }
+
+    bytesToSend -= bytesSent;
+    currentBytes += bytesSent;
+  }
+}
+
+template<typename T>
+void recvBytes(int socket, T* buffer, std::size_t length) {
+
+  size_t bytesToReceive = sizeof(T) * length;
+  if (bytesToReceive == 0) {
+    return;
+  }
+
+  auto bytes = reinterpret_cast<uint8_t*>(buffer);
+  uint8_t *currentBytes = bytes;
+
+  while (bytesToReceive > 0) {
+    ssize_t bytesReceived;
+    SYSCHECK(bytesReceived = ::recv(socket, currentBytes, bytesToReceive, 0))
+    if (bytesReceived == 0) {
+      throw std::system_error(ECONNRESET, std::system_category());
+    }
+
+    bytesToReceive -= bytesReceived;
+    currentBytes += bytesReceived;
+  }
+}
+
+// send a vector's length and data
+template<typename T>
+void sendVector(int socket,
+                const std::vector<T>& vec,
+                bool moreData = false) {
+
+  SizeType size = vec.size();
+  sendBytes<SizeType>(socket, &size, 1, true);
+  sendBytes<T>(socket, vec.data(), size, moreData);
+}
+
+// receive a vector as sent in sendVector
+template<typename T>
+std::vector<T> recvVector(int socket) {
+  SizeType valueSize;
+  recvBytes<SizeType>(socket, &valueSize, 1);
+  std::vector<T> value(valueSize);
+  recvBytes<T>(socket, value.data(), value.size());
+  return value;
+}
+
+// this is only for convenience when sending rvalues
+template<typename T>
+void sendValue(int socket, const T& value, bool moreData = false) {
+  sendBytes<T>(socket, &value, 1, moreData);
+}
+
+template<typename T>
+T recvValue(int socket) {
+  T value;
+  recvBytes<T>(socket, &value, 1);
+  return value;
+}
+
+// send a string's length and data
+inline void sendString(int socket,
```

Comment:
At some point we should just move to an overloaded `send` and `recv` function. They're really verbose at the moment.

Reply:
Yeah :)